drivers/ad7606/ad7606.c

```diff
@@ -0,0 +1,220 @@
+{
+	uint8_t size;
+	int32_t ret;
+
+	size = ad7606_chip_info_tbl[dev->device_id].num_channels * 2;
+	ret = gpio_set_value(dev->gpio_convst, 0);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_set_value(dev->gpio_convst, 1);
+	if (ret < 0)
+		return ret;
+
+	return spi_write_and_read(dev->spi_desc, dev->data, size);
+}
+
+int32_t ad7606_spi_read_samples(struct ad7606_dev *dev,
+				uint8_t channel,
+				uint16_t *adc_data)
+{
+	int32_t ret;
+
+	ret = ad7606_spi_read_bulk(dev);
+	if (ret < 0)
+		return ret;
+
+	*adc_data = (dev->data[channel * 2] << 8 | dev->data[channel * 2 + 1]);
+
+	return 0;
+}
+
+int32_t ad7606_reset(struct ad7606_dev *dev)
+{
+	int32_t ret;
+
+	ret = gpio_set_value(dev->gpio_reset, 1);
+	if (ret < 0)
+		return ret;
+
+	mdelay(1);
+
+	return gpio_set_value(dev->gpio_reset, 0);
+}
+
+int32_t ad7606_request_gpios(struct ad7606_dev *dev,
+			     struct ad7606_init_param *init_param)
+{
+	int32_t ret;
+
+	ret = gpio_get(&dev->gpio_reset, init_param->gpio_reset);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_convst, init_param->gpio_convst);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_busy, init_param->gpio_busy);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_range, init_param->gpio_range);
+	if (ret < 0)
+		return ret;
+
+	if (!ad7606_chip_info_tbl[dev->device_id].has_oversampling)
+		return 0;
+
+	ret = gpio_get(&dev->gpio_range, init_param->gpio_os0);
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_get(&dev->gpio_range, init_param->gpio_os1);
+	if (ret < 0)
+		return ret;
+
+	return gpio_get(&dev->gpio_range, init_param->gpio_os2);
+}
+
+int32_t ad7606_set_os_ratio(struct ad7606_dev *dev,
+			    enum ad7606_osr osr)
+{
+	int32_t ret;
+
+	ret = gpio_set_value(dev->gpio_os0, ((osr & 0x01) >> 0));
+	if (ret < 0)
+		return ret;
+
+	ret = gpio_set_value(dev->gpio_os1, ((osr & 0x02) >> 1));
+	if (ret < 0)
+		return ret;
+
+	return gpio_set_value(dev->gpio_os2, ((osr & 0x04) >> 2));
+}
+
+int32_t ad7606_init(struct ad7606_dev **device,
+		    struct ad7606_init_param *init_param)
+{
+	struct ad7606_dev *dev;
+	int32_t ret;
+
+	dev = (struct ad7606_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -1;
+
+	ret = spi_init(&dev->spi_desc, &init_param->spi_init);
+	if (ret < 0)
+		goto error;
+
+	dev->device_id = init_param->device_id;
+
+	ret = ad7606_request_gpios(dev, init_param);
+	if (ret < 0)
+		goto error;
+
+	ret = ad7606_reset(dev);
+	if (ret < 0)
+		goto error;
+
+	ret = gpio_set_value(dev->gpio_range, init_param->range);
+	if (ret < 0)
+		goto error;
+
+	ret = gpio_set_value(dev->gpio_convst, 1);
+	if (ret < 0)
+		goto error;
+
+	if (ad7606_chip_info_tbl[dev->device_id].has_oversampling)
+		ad7606_set_os_ratio(dev, init_param->osr);
+
+	*device = dev;
+
+	printf("ad7606 successfully initialized\n");
+
+	return 0;
+error:
+	printf("ad7606 initialization failed\n");
+	ad7606_remove(dev);
+	return ret;
+}
+
+int32_t ad7606_remove(struct ad7606_dev *dev)
```

Comment:
the naming here would need a bit of work
something like `ad7606_destroy()` or `ad7606_uninit()` ; the `remove` term is more Linux-y

Reply:
This just follows the convention that we agreed upon in `platform_drivers`. i.e. `gpio_remove()`, `spi_remove()`

Reply:
ack
let's leave it like this